Packages/com.unity.inputsystem/InputSystem/Plugins/OnScreen/OnScreenControl.cs

```diff
@@ -25,18 +27,140 @@ namespace UnityEngine.Experimental.Input.Plugins.OnScreen
     /// </remarks>
     public abstract class OnScreenControl : MonoBehaviour
     {
-        [SerializeField] private string m_ControlPath;
+        private struct OnScreenDeviceEventInfo
+        {
+            public InputEventPtr eventPtr;
+            public NativeArray<byte> buffer;
+            public InputDevice device;
+        }
+
+        private static InlinedArray<OnScreenDeviceEventInfo> s_DeviceEventInfoArray = new InlinedArray<OnScreenDeviceEventInfo>();
+
+        // We will have N devices mapped to X number of OnScreenControls
+        // Need to keep track and reference count how many controls are sharing
+        // the same device, so that when the last control is ready to be destroyed
+        // it can clean up the memory and remove the device from the input system.
+        private static List<int> s_RegisteredOnScreenControls = new List<int>();
+
+        public string controlPath
+        {
+            get { return m_ControlPath; }
+            set
+            {
+                m_ControlPath = value;
+                if (enabled)
+                {
+                    SetupInputControl();
+                }
+            }
+        }
+
+        [NonSerialized] internal InputControl m_Control;
+        [SerializeField] internal string m_ControlPath;
+
+        private int m_DeviceEventInfoIndex;
+        private string m_Layout;
+
+        private static int GetDeviceEventIndex(string layout)
+        {
+            for (int index = 0; index < s_DeviceEventInfoArray.length; index++)
+            {
+                if (s_DeviceEventInfoArray[index].device.layout == layout)
+                    return index;
+            }
+            return -1;
+        }
+
+        private static int CreateOnScreenDevice(string layout)
+        {
+            var device = InputSystem.AddDevice(layout);
+            InputEventPtr eventPtr;
+            var buffer = StateEvent.From(device, out eventPtr, Allocator.Persistent);
+
+            // Need to cache the buffer, device and InputEventPointer
+            // so that muliple OnScreenControlInstances can share
+            // the same OnScreenDevice and input memory;
+            OnScreenDeviceEventInfo deviceEventInfo;
+            deviceEventInfo.eventPtr = eventPtr;
+            deviceEventInfo.buffer = buffer;
+            deviceEventInfo.device = device;
+
+            s_DeviceEventInfoArray.Append(deviceEventInfo);
+
+            // Give the caller the index in the cache.
+            return s_DeviceEventInfoArray.length - 1;
+        }
+
+        private static void RemoveOnScreenDevice(int id)
+        {
+            s_DeviceEventInfoArray[id].buffer.Dispose();
+            InputSystem.RemoveDevice(s_DeviceEventInfoArray[id].device);
```

Comment:
You're not actually removing the entry from s_DeviceEventInfoArray. This is why you're crashing. `buffer` remains initialized with the data you have disposed and the next test writes into the now invalid memory.